### PR TITLE
:sparkles: feat: add offline mode

### PR DESCRIPTION
**Pull request by rcattin**
# Description
add an option to download a github repo
improve model repo mode choice in cli

Fixes THipster/THipster#94

## Type of change

:sparkles: New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

test list & use

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by rcattin** [comment]
Related model repo changes
https://github.com/THipster/models/compare/9f96c371d67d35ed0a500b35e6a2c1435df02d48...22ef50418e9127212d9ae0638a9f6675044f4dfc

File: tests/test_repository.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+@pytest.fixture
+def create_models_directory():
+    models_path = os.path.join(get_app_dir(constants.APP_NAME), 'models')
+    os.mkdir(models_path)
+
+    yield models_path
+
+    if os.path.exists(models_path):
+        os.rmdir(models_path)
```
Comment (gsuquet) [inline tests/test_repository.py:21]: https://beta.ruff.rs/docs/rules/#flake8-use-pathlib-pth
Replace:  
- os.path.join with Path(get_app_dir(constants.APP_NAME)) / 'models'
- os.mkdir(models_path) with Path(models_path).mkdir()
- os.rmdir(models_path) with Path(models_path).rmdir()

File: tests/test_repository.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+
+@pytest.fixture
+def create_example_repo(create_models_directory):
+    repo_path = os.path.join(create_models_directory, 'example')
+    os.mkdir(repo_path)
+
+    yield 'example'
+
+    os.rmdir(repo_path)
```
Comment (gsuquet) [inline tests/test_repository.py:31]: Same as above

File: thipstercli/cli.py [hunk 3]
```diff
@@ -165,5 +157,53 @@ def _run(
         error(*e.args)
 
 
+def get_repo(local_path, online_path, branch):
+    if local_path:
+        if local_path in repository.list_installed_repos():
+            repo_path = os.path.join(
```
Comment (gsuquet) [inline thipstercli/cli.py:163]: Same replace os.path with Path()

File: thipstercli/cli.py [hunk 4]
```diff
@@ -165,5 +157,53 @@ def _run(
...
+        GithubRepo(
+            online_path,
+            branch,
+        )
+    else:
+        match state.get('repository_recovery_mode'):
+            case 'local':
+                repo_path = os.path.join(
+                    config.app_dir, 'models',
+                    state.get('models_repository'),
```
Comment (gsuquet) [inline thipstercli/cli.py:190]: same https://beta.ruff.rs/docs/rules/#flake8-use-pathlib-pth

File: thipstercli/repository.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+    repositories_path = os.path.join(app_dir, LOCAL_MODELS_REPOSITORY_PATH)
+    if not os.path.exists(repositories_path):
+        os.mkdir(repositories_path)
+
+    clone_to = os.path.join('/tmp', 'thipster')
+    Repo.clone_from(url, clone_to)
+
+    with open(os.path.join(clone_to, 'thipster-config.json')) as f:
+        config_path = f.read()
```
Comment (gsuquet) [inline thipstercli/repository.py:79]: pathlib Path()

File: thipstercli/repository.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        error('Configuration error in thipster-config.json')
+
+    if os.path.exists(repo_name):
+        shutil.rmtree(clone_to)
+        error(f'Repository named {repo_name} already installed')
+
+    try:
+        shutil.move(
+            os.path.join(clone_to, repo_directory),
+            os.path.join(repositories_path, repo_name),
```
Comment (gsuquet) [inline thipstercli/repository.py:101]: pathlib Path()

File: thipstercli/repository.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        shutil.rmtree(clone_to)
+        error('Configuration error in thipster-config.json : bad model_folder')
+
+    shutil.rmtree(clone_to)
+
+
+def list_installed_repos():
+    downloaded_repos = []
+    repositories_path = os.path.join(app_dir, LOCAL_MODELS_REPOSITORY_PATH)
+    if os.path.exists(repositories_path):
```
Comment (gsuquet) [inline thipstercli/repository.py:113]: Pathlib Path()

File: tests/test_repository.py (hunk 1 above)
Comment (rcattin) [reply tests/test_repository.py]: Rule wasn't in ruff config btw, just added it

**Review by gsuquet — APPROVED** [review]

**Pull request merged**